mypy/checkexpr.py

```diff
@@ -2122,15 +2098,122 @@ def infer_function_type_arguments(
                 else:
                     pass1_args.append(arg)
 
-            inferred_args, _ = infer_function_type_arguments(
-                callee_type,
-                pass1_args,
-                arg_kinds,
-                arg_names,
-                formal_to_actual,
-                context=self.argument_infer_context(),
-                strict=self.chk.in_checked_function(),
-            )
+            if True:  # NEW CODE
+                # compute the inner constraints
+                _inner_constraints = infer_constraints_for_callable(
+                    callee_type,
+                    pass1_args,
+                    arg_kinds,
+                    arg_names,
+                    formal_to_actual,
+                    context=self.argument_infer_context(),
+                )
+                # HACK: convert "Literal?" constraints to their non-literal versions.
+                inner_constraints: list[Constraint] = []
+                for constraint in _inner_constraints:
+                    target = get_proper_type(constraint.target)
+                    inner_constraints.append(
+                        Constraint(
+                            constraint.original_type_var,
+                            constraint.op,
+                            (
+                                target.copy_modified(last_known_value=None)
+                                if isinstance(target, Instance)
+                                else target
+                            ),
+                        )
+                    )
```

Comment:
This hack is needed for `testLiteralAndGenericWithUnion` and [`testLiteralMappingContext`](https://github.com/python/mypy/pull/19399/commits/76f357154c0d0f6b6e6fb649b8da2aba5e3eac2a).

The only purpose is to convert constraints like `T ≷ Literal['foo"]?` to `T ≷ str`, which causes the joint solution to fail due to incompatible constraints `T <: Literal["foo"]` and `T :> str`, hence the longer route of using outer solution first is used.